pkg/remotewrite/remotewrite.go

```diff
@@ -89,76 +88,225 @@ func (o *Output) flush() {
-	promTimeSeries := make([]prompb.TimeSeries, 0)
+func (o *Output) convertToPbSeries(samplesContainers []metrics.SampleContainer) []prompb.TimeSeries {
+	// The seen map is required because the samples containers
+	// could have several samples for the same time series
+	//  in this way, we can aggregate and flush them in a unique value
+	//  without overloading the remote write endpoint.
+	//
+	// It is also essential because the core generates timestamps
+	// with a higher precision (ns) than Prometheus (ms),
+	// so we need to aggregate all the samples in the same time bucket.
+	// More context can be found in the issue
+	// https://github.com/grafana/xk6-output-prometheus-remote/issues/11
+	seen := make(map[string]struct{})
 
 	for _, samplesContainer := range samplesContainers {
 		samples := samplesContainer.GetSamples()
 
 		for _, sample := range samples {
-			// Prometheus remote write treats each label array in TimeSeries as the same
-			// for all Samples in those TimeSeries (https://github.com/prometheus/prometheus/blob/03d084f8629477907cab39fc3d314b375eeac010/storage/remote/write_handler.go#L75).
-			// But K6 metrics can have different tags per each Sample so in order not to
-			// lose info in tags or assign tags wrongly, let's store each Sample in a different TimeSeries, for now.
-			// This approach also allows to avoid hard to replicate issues with duplicate timestamps.
-
-			labels, err := tagsToLabels(sample.Tags, o.config)
-			if err != nil {
-				o.logger.Error(err)
-			}
-
-			if newts, err := o.metrics.transform(o.mapping, sample, labels); err != nil {
-				o.logger.Error(err)
+			truncTime := sample.Time.Truncate(time.Millisecond)
+			timeSeriesKey := timeSeriesKey(sample.Metric, sample.Tags)
+			swm, ok := o.tsdb[timeSeriesKey]
+			if !ok {
+				swm = &seriesWithMeasure{
+					TimeSeries: TimeSeries{
+						Metric: sample.Metric,
+						Tags:   sample.Tags,
+					},
+					Measure: sinkByType(sample.Metric.Type),
+					Latest:  truncTime,
+				}
+				o.tsdb[timeSeriesKey] = swm
+				seen[timeSeriesKey] = struct{}{}
 			} else {
-				promTimeSeries = append(promTimeSeries, newts...)
+				// save as a seen item only when the samples have a time greater than
+				// the previous saved, otherwise some implementations
+				// could see it as a duplicate and generate warnings (e.g. Mimir)
+				if truncTime.After(swm.Latest) {
+					swm.Latest = truncTime
+					seen[timeSeriesKey] = struct{}{}
+				}
+
+				// If current == previous:
+				// the current received time before being truncated had a higher precision.
+				// It's fine to aggregate them but we avoid to add to the seen map because:
+				// - in the case it is a new flush operation then we avoid delivering
+				//   for not generating duplicates
+				// - in the case it is in the same operation but across sample containers
+				//   then the time series should be already on the seen map and we can skip
+				//   to re-add it.
```

Comment:
@codebien When does this case actually happen? The same metric, out of order (i.e. with `ts2 <= ts1`) and in different sample containers?

I wonder if it shouldn't be a warning / error level as well.